### PR TITLE
constraints

### PR DESCRIPTION
**Pull request by sonichi**
generalize the mem constraints to generic constraints

### CODE REVIEW

File: test/tune/test_tune.py [hunk 1]
```diff
@@ -163,21 +166,28 @@ def test_nested():
     }
 
     def simple_func(config):
-        tune.report(metric=(config["cost_related"]["a"] - 4)**2
-                    * (config["b"] - 0.7)**2)
+        metric = (config["cost_related"]["a"] - 4)**2 \
```
Comment (qingyun-wu) [inline test/tune/test_tune.py:169]: change 'metric' to 'obj'

File: flaml/version.py [hunk 2]
```diff
@@ -1 +1 @@
-__version__ = "0.3.6"
```
Comment (qingyun-wu) [inline flaml/version.py:1]: do not change version

File: flaml/searcher/blendsearch.py [hunk 3]
```diff
@@ -39,9 +39,11 @@ def __init__(self,
                  min_resource: Optional[float] = None,
                  max_resource: Optional[float] = None,
                  reduction_factor: Optional[float] = None,
-                 resources_per_trial: Optional[dict] = None,
                  global_search_alg: Optional[Searcher] = None,
-                 mem_size: Callable[[dict], float] = None,
+                 config_constraints: Optional[
+                     List[Tuple[Callable[[dict], float], str, float]]] = None,
```
[the remark this answers is not in the document]
Comment (sonichi) [reply flaml/searcher/blendsearch.py:44]: list of arguments

File: flaml/version.py (hunk 2 above)
Comment (sonichi) [reply flaml/version.py:1]: changing version number doesn't requires publishing to pypi.

File: test/tune/test_tune.py (hunk 1 above)
Comment (sonichi) [reply test/tune/test_tune.py]: done

File: flaml/searcher/blendsearch.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (aminsaied) [reply flaml/searcher/blendsearch.py:44]: This is hard to parse: List[Tuple[Callable[[dict], float], str, float]]]

What about introducing a lightweight dataclass to manage these? e.g.

@dataclass
class Constraint:
  metric_function: Callable[Dict, float]
  inequality: str
  value: float

and replace type with Optional[List[Constraint]]. This would also make the code more readable below.

**Review by qingyun-wu — APPROVED** [review]
